_posts/2018-03-21-building-a-model-for-atomic-event-data-as-a-graph/index.md

```diff
@@ -0,0 +1,151 @@
+
+Graph databases are also designed for low-latency performance. That makes them a better choice than relational databases for certain applications, such as recommendation engines, especially at scale.
+
+There are some questions (for example path analysis) that are difficult to answer when using a relational database, but that are easy to answer with a graph. (We've [experimented][link1] with path analysis in Neo4j before.)
+
+The schema-free (as in the way the data fits together in the DB, not as in 'JSON schema'-free) nature of graph databases means graphs are readily extensible through the addition of new graphs. For example, you could add geolocation data modelled as a graph and link it to your atomic data by relationships that do not otherwise mess with the atomic data model. That would open up a whole new avenue for enriching the data in-database.
+
+Adding support for a graph storage target then makes a lot of sense. But while with RDBMSs the core data model is the same and we only have to account for the way different vendors implement that model, with graph databases there are some crucial questions that are as yet open.
+
+## How do we model event data as a graph?
+
+In the Snowplow event data model (as currently implemented in RDBMSs), each event is a row in a table. The table has as many columns as there are facets to that event, eg user, timestamp, URL, etc. The data may be shredded into separate tables, some of which may have a many-to-one relationship with the atomic data. But the underlying model is the same. There really isn't much scope for deviation from it.
+
+In graph databases, there are many ways event data can be modelled. One example would be to make all the entities in the event nodes in the graph, and then draw relationships between those nodes. Another would be to make the event itself a node that has outgoing relationships to its various attributes, such as user, timestamp, URL, etc. And then there are combinations and variations of these two basic models. It's not obvious which one of these models is 'the right one', if there even is such a thing.
+
+Graphs excel at modelling networks -- social, communication, supply, etc. You can even model and link different types of networks in an overarching graph, which is very powerful. In that environment, events would be best represented by relationships between entities in those networks.
+
+Since graphs are so flexible, the advice is usually to build a model that suits the queries you want to run. Snowplow users come from all sorts of industries and are trying to solve a great variety of challenges. Whatever our model, it needs to be able to accommodate that diversity.
+
+### Event grammar
+
+We've already done great [work][link2] on developing a generalised event grammar that goes some way towards a graph data model for event data. In that model, an event is a snapshot of set of entities in time, eg a subject, an object, context, etc.
```

Comment:
I agree we have done some great work here, but I'd rephrase this a little to not sound like we're congratulating ourselves :)